scio-core/src/main/scala/com/spotify/scio/io/dynamic/package.scala

```diff
@@ -0,0 +1,151 @@
+ * }}}
+ */
+package object dynamic {
+
+  /**
+   * Enhanced version of [[com.spotify.scio.values.SCollection SCollection]] with dynamic
+   * destinations methods.
+   */
+  implicit class DynamicTextIoSCollection(val self: SCollection[String]) extends AnyVal {
+    /**
+     * Save this SCollection as text files specified by the
+     * [[org.apache.beam.sdk.io.FileBasedSink.DynamicDestinations DynamicDestinations]].
+     */
+    def saveAsTextFile(destinations: DynamicDestinations[String, _, String])
+    : Future[Tap[String]] = {
+      if (self.context.isTest) {
+        throw new NotImplementedError(
+          "Text file with dynamic destinations cannot be used in a test context")
+      } else {
+        val tempDir = FileSystems.matchNewResource(self.context.options.getTempLocation, true)
+        self.applyInternal(TextIO.write().to(destinations).withTempDirectory(tempDir))
+      }
+
+      Future.failed(
+        new NotImplementedError("Text file future with dynamic destinations not implemented"))
+    }
+
+    /**
+     * Save this SCollection as text files specified by the destination function.
+     */
+    def saveAsTextFile(path: String, suffix: String = ".txt", windowedWrites: Boolean = false,
+                       defaultDestination: String = "default")
+                      (destinationFn: String => String)
+    : Future[Tap[String]] = {
+      val destinations = newDynamicDestination(
+        path, suffix, windowedWrites, defaultDestination, destinationFn)
+      saveAsTextFile(destinations)
+    }
+  }
+
+  /**
+   * Enhanced version of [[com.spotify.scio.values.SCollection SCollection]] with dynamic
+   * destinations methods.
+   */
+  implicit class DynamicIoSCollection[T](val self: SCollection[T]) extends AnyVal {
+    /**
+     * Save this SCollection as Avro files specified by the
+     * [[org.apache.beam.sdk.io.DynamicAvroDestinations DynamicAvroDestinations]].
+     */
+    def saveAsAvroFile(destinations: DynamicAvroDestinations[T, _, T], schema: Schema = null)
+    : Future[Tap[T]] = {
+      if (self.context.isTest) {
+        throw new NotImplementedError(
+          "Avro file with dynamic destinations cannot be used in a test context")
+      } else {
+        val tempDir = FileSystems.matchNewResource(self.context.options.getTempLocation, true)
+        val cls = self.ct.runtimeClass.asInstanceOf[Class[T]]
+        val t = if (classOf[SpecificRecordBase] isAssignableFrom cls) {
+          AvroIO.write(cls)
+        } else {
+          AvroIO.writeGenericRecords(schema).asInstanceOf[AvroIO.Write[T]]
+        }
+        self.applyInternal(t.to(destinations).withTempDirectory(tempDir))
+      }
+
+      Future.failed(
+        new NotImplementedError("Avro file future with dynamic destinations not implemented"))
+    }
+
+    /**
+     * Save this SCollection as Avro files specified by the destination function.
+     */
+    def saveAsTextFile(path: String, schema: Schema = null,
```

Comment:
@nevillelyh should this be `saveAsAvroFile`?

Reply:
Duh! 🤯